File: Algorithms/Graph/Flood Fill/Flood Fill.cpp
typedef pair<int, int> ii;

const int MAX_R = XXX, MAX_C = XXX;

int R, C;  // define in main
int grid[MAX_R][MAX_C];
int dc[] = {-1, 0, 1, -1, 1, -1, 0, 1};
int dr[] = {-1, -1, -1, 0, 0, 1, 1, 1};

void flood_fill_DFS(int row, int col) {
  if (row < 0 || row >= R || col < 0 || col >= C || grid[row][col] == 1)
    return;
  grid[row][col] = 1;

  for (int i = 0; i < 8; i++)
    flood_fill_DFS(row + dr[i], col + dc[i]);
}

void flood_fill_BFS(int r0, int c0) {
  queue<ii> q;

  int col, row, r, c;
  q.push(make_pair(r0, c0));

  while (!q.empty()) {
    row = q.front().first;
    col = q.front().second;
    q.pop();

    if (row < 0 || row >= R || col < 0 || col >= C || grid[row][col] == 1)
      continue;

    grid[row][col] = 1;
    for (int i = 0; i < 8; i++) {
      r = row + dr[i];
      c = col + dc[i];
      q.push(make_pair(r, c));
    }
  }
}

File: Algorithms/Graph/Flood Fill/Flood Fill.py
from collections import deque

# R, C = XXX, XXX  # define in main
# define grid in main
dc = [-1, 0, 1, -1, 1, -1, 0, 1]
dr = [-1, -1, -1, 0, 0, 1, 1, 1]


def flood_fill_DFS(row, col):
    if row < 0 or row >= R or col < 0 or col >= C or grid[row][col] == 1:
        return
    grid[row][col] = 1

    for c, r in zip(dc, dr):
        flood_fill_DFS(row + r, col + c)


def flood_fill_BFS(r0, c0):
    q = deque()
    q.append((r0, c0))

    while q:
        row, col = q.popleft()
        if row < 0 or row >= R or col < 0 or col >= C or grid[row][col] == 1:
            continue
        grid[row][col] = 1

        for i, j in zip(dc, dr):
            r = row + j
            c = col + i
            q.append((r, c))
